File: 10-DesignPatterns-CSharp/10-2-2-Adapter/Ducks/Duck.cs
namespace Adapter.Ducks
{
    using System;
    using Adapter.Fly;
    using Adapter.Quack;

    public abstract class Duck
    {
        private readonly IQuackStrategy _quackStrategy;
        private readonly IFlyStrategy _flyStrategy;

        protected Duck(
            IQuackStrategy quackStrategy,
            IFlyStrategy flyStrategy)
        {
            _quackStrategy = quackStrategy;
            _flyStrategy = flyStrategy;
        }

        public void Quack() => _quackStrategy.Quack();

        public void Swim() => Console.WriteLine("Swimming.");

        public void Fly() => _flyStrategy.Fly();

        public abstract void Display();
    }
}

File: 10-DesignPatterns-Java/Starter/Adapter/src/Ducks/Duck.java
package Ducks;

import Fly.IFlyStrategy;
import Quack.IQuackStrategy;

public abstract class Duck {
    private final IQuackStrategy _quackStrategy;
    private final IFlyStrategy _flyStrategy;

    protected Duck(IQuackStrategy quackStrategy, IFlyStrategy flyStrategy)
    {
        _quackStrategy = quackStrategy;
        _flyStrategy = flyStrategy;
    }

    public void quack() {
        _quackStrategy.quack();
    }

    public void swim() {
        System.out.println("Swimming.");
    }

    public void fly() {
        _flyStrategy.fly();
    }

    public abstract void display();
}
